src/swarm/Swarm.ts

```diff
@@ -166,70 +187,107 @@ export const updateUserFeed = async (feedTemplate: FeedTemplate, identity: Priva
+    downloadFeed: async (feedUri: string, timeout: number = 0) => await downloadFeed(swarmGateway, feedUri, timeout),
+    getUri: () => `bzz-feed:/?${calculateFeedAddressQueryString(address)}`,
+    address,
+});
+
+export type FeedDigestSigner = (digest: number[]) => string | Promise<string>;
+
+export interface WriteableFeedApi extends ReadableFeedApi {
     updateWithFeedTemplate: (feedTemplate: FeedTemplate, data: string) => Promise<FeedTemplate>;
     update: (data: string) => Promise<FeedTemplate>;
-    downloadFeed: (feedUri: string) => Promise<string>;
-    getUri: () => string;
+    signFeedDigest: FeedDigestSigner;
 }
 
-export const makeFeedApi = (identity: PrivateIdentity): FeedApi => {
+export const makeFeedApi = (address: FeedAddress, signFeedDigest: FeedDigestSigner, swarmGateway: string = DefaultGateway): WriteableFeedApi => {
     return {
-        download: async (): Promise<string> => downloadUserFeed(identity),
-        downloadPreviousVersion: async (epoch: Epoch) => downloadUserFeedPreviousVersion(identity.address, epoch),
-        downloadFeedTemplate: async () => downloadUserFeedTemplate(identity),
-        updateWithFeedTemplate: async (feedTemplate: FeedTemplate, data) => await updateUserFeed(feedTemplate, identity, data),
+        ...makeReadableFeedApi(address, swarmGateway),
+
+        updateWithFeedTemplate: async (feedTemplate: FeedTemplate, data) => await updateUserFeedWithSignFunction(swarmGateway, feedTemplate, signFeedDigest, data),
         update: async (data: string): Promise<FeedTemplate> => {
-            const feedTemplate = await downloadUserFeedTemplate(identity);
-            return await updateUserFeed(feedTemplate, identity, data);
+            const feedTemplate = await downloadUserFeedTemplate(swarmGateway, address);
+            return await updateUserFeedWithSignFunction(swarmGateway, feedTemplate, signFeedDigest, data);
         },
-        downloadFeed: async (feedUri: string) => await downloadFeed(feedUri),
-        getUri: () => `bzz-feed:/?user=${identity.address}`,
+        signFeedDigest,
     };
 };
 
 export interface BzzApi {
-    download: (hash: string) => Promise<string>;
+    download: (hash: string, timeout: number) => Promise<string>;
     upload: (data: string) => Promise<string>;
     uploadPhoto: (localPath: string) => Promise<string>;
 }
 
 export const makeBzzApi = (swarmGateway: string = DefaultGateway): BzzApi => {
     return {
-        download: (hash: string) => download(hash, swarmGateway),
+        download: (hash: string, timeout: number = 0) => downloadData(hash, timeout, swarmGateway),
         upload: (data: string) => upload(data, swarmGateway),
         uploadPhoto: (localPath: string) => uploadPhoto(localPath, swarmGateway),
     };
 };
 
-export interface Api {
+export interface BaseApi {
+    readonly swarmGateway: string;
+}
+
+export interface WriteableApi extends BaseApi {
```

Comment:
Reader / Writer might be better.

Reply:
I don't like Readable / Writeable either, right now my best idea is `DownloadApi`, `UploadApi`, `FeedDownloadApi` with `FeedUpdateApi`)

Reply:
I'm good with Download(er) too.